### DISCUSSION TITLE
Handling form validations in an API route

### DISCUSSION THREAD

**Discussion by lodybo** [vercel/next.js#40257] [Help] [answered]
I was wondering if it's possible to retrieve the data from a handler in response to a form submit. Usually this is done through intercepting the form submit and perform a `fetch` request, but in the event JS isn't loaded the form would simply redirect to the handler page.

What I'd like to do is the following, given a simple form, like a login page with a username and password field:
- Create a route with a `<form method="post">{ ... }</form>` that has a username and password field.
- Create a handler at `/api/login.js` that validates the form data and checks whether the login succeeds.
- If form validation fails, return a JSON response with the specific error information for the user. The form route gets this information and marks the corresponding inputs as invalid and shows the error **<== this is what I'd like to achieve**.
- If the form data is filled in but the login check fails (because of incorrect credentials), return a JSON response with a notice for the user, prompting them to check their credentials **<== this is a similar usecase to the one above**.
- If the login check succeeds, redirect the user to a different page.

Demonstrated in code, it would look a bit like this:
```jsx

// /api/login.js
export async function handler(req, res) {
  const { username, password } = req.body;
  
  const errors = {
    // validateUsername and validatePassword return a string with the error message or undefined.
    username: validateUsername(username),
    password: validatePassword(password),
  };
  
  if (errors.username || errors.password) {
    return res.status(402).json(errors); // Point 1
  }
  
  const validCredentials = await checkCredentials(username, password);
  
  if (!validCredentials) {
    return res.status(401).json({ credentials: 'Username and/or password is incorrect.' }); // Point 2
  }
  
  return res.redirect('/profile');
}

// pages/login.jsx
export default function LoginPage() {
  // How do I access the responses that might be returned at points 1 and 2 in the handler function?
  
  return (
    <form method="post">
      <label>
	<p>Username:</p>
	<input type="text" name="name" />
      </label>
			
      <label>
	<p>Password:</p>
	<input type="password" name="password" />
      </label>
			
      <button type="submit">Log in</button>
    </form>
  );
}
```

**Comment by icyJoseph** [comment, marked as answer]
Hi,

The issue here is that when the form is submitted the browser attempts to navigate to the action path. AFAIK, the only way to prevent that is with JavaScript, and as you say, if JS is disabled, then we seem to be in a bit of a problem. 

> Note that typically, navigating the document unloads the entirety of the page.

This is where a project called `next-runtime`, https://next-runtime.meijer.ws/getting-started/1-introduction, comes in handy:

```tsx
import type { NextPage } from "next";

import { handle, json } from "next-runtime";

export const getServerSideProps = handle<
  { message: string } | { name: string }
>({
  async get({ params, query }) {
    return json({ name: "foo" });
  },

  async post({ req: { body } }) {
    console.log(body);
    return json({ message: "Thanks for your submission!" });
  },
});

const Home: NextPage<
  { message: string; name: never } | { name: string; message: never }
> = ({ message }) => {
  if (message) return <div>{message}</div>;

  return (
    <form method="post">
      <label>
        <p>Username:</p>
        <input type="text" name="username" />
      </label>

      <label>
        <p>Password:</p>
        <input type="password" name="password" />
      </label>

      <button type="submit">Log in</button>
    </form>
  );
};

export default Home;
```

You see we are able to post to the same route, where `getServerSideProps` is used to catch all types of requests by method. 

> One caveat is that even this method does document navigation, meaning that the page is destroyed and recreated by the browser. In other words, any state you had, any allocated objects, etc, are all discarded, as if you were just arriving to the site.

In fact, we can use this to extend our knowledge of `getServerSideProps`, let's remove the additional library and roll our own solution. Not saying that we should roll our own, but we should understand what's going on.

```tsx
import type { GetServerSidePropsContext, NextPage } from "next";

export const getServerSideProps = async (ctx: GetServerSidePropsContext) => {
  if (ctx.req.method === "POST") {
    const chunks = [];

    for await (const chunk of ctx.req) {
      chunks.push(chunk);
    }

    const data = Buffer.concat(chunks).toString();

    const asJSON = data
      .split("&")
      .reduce<Record<string, string>>((prev, curr) => {
        const [key, value] = curr.split("=");
        prev[key] = value;
        return prev;
      }, {});

    console.log({ asJSON });

    return { props: { message: "hello world" } };
  }

  return { props: { name: "foo" } };
};

const Home: NextPage<
  { message: string; name: never } | { name: string; message: never }
> = ({ message }) => {
  if (message) return <div>{message}</div>;

  return (
    <form method="post">
      <label>
        <p>Username:</p>
        <input type="text" name="username" />
      </label>

      <label>
        <p>Password:</p>
        <input type="password" name="password" />
      </label>

      <button type="submit">Log in</button>
    </form>
  );
};

export default Home;
```

Downsides of rolling your own is that there's edge cases, and situations where parsing the body of request should be done by an specialised library. In general, you should reduce the amount of responsibility the code you write has, versus how much you place on the packages you use. In this case, it is fair to offload the responsibility of dealing with parsing the body and detecting the method used to `next-runtime`, but the second example helps us understand what might be going on under the hood.

**Reply by lodybo** [reply]
Thanks @icyJoseph! I'd found Stephan's blog but hadn't checked out the `next-runtime` package. After reading the documentation it covers the use cases that I want.

And, like him, I was initially inspired by the way forms work in Remix. Happy to have a similar workflow here for Next.js!
